### PR TITLE
Fix search highlight for non-unicode chars

### PR DESCRIPTION
**Pull request by Samyak2**
# Pull Request

## What does this PR do?
Fixes https://github.com/meilisearch/MeiliSearch/issues/1480

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

## Changes

The `matching_bytes` function takes a `&Token` now and:
- gets the number of bytes to highlight (unchanged).
- uses `Token.num_graphemes_from_bytes` to get the number of grapheme clusters to highlight.

In essence, the `matching_bytes` function now returns the number of matching grapheme clusters instead of bytes.

Added proper highlighting in the HTTP UI:
- requires dependency on `unicode-segmentation` to extract grapheme clusters from tokens
- `<mark>` tag is put around only the matched part
    - before this change, the entire word was highlighted even if only a part of it matched

## Questions

Since `matching_bytes` does not return number of bytes but grapheme clusters, should it be renamed to something like `matching_chars` or `matching_graphemes`? Will this break the API?

Thank you very much @ManyTheFish for helping :smile:

### CODE REVIEW

**Comment by Samyak2** [comment]
Ah I just found out that the tests for `milli` fail. Will fix in some time.

**Comment by Samyak2** [comment]
I have fixed the tests to build successfully, but they still fail. I think it's due a mistake I made in https://github.com/meilisearch/tokenizer/pull/59/files#diff-76ada62566331039fde44902c7d08ec1f2c977a147cdefcefc466e07cedafc4fR89

It should be:
```rust
            None => num_bytes,
```

**Comment by Samyak2** [comment]
Opened a PR there: https://github.com/meilisearch/tokenizer/pull/64

Sorry about this :grimacing:

**Comment by curquiza** [comment]
Thanks @Samyak2 for your work 😄 

@ManyTheFish @Kerollmops https://github.com/meilisearch/tokenizer/pull/64 should be reviewed (and merged if accepted) before merging this PR

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

**Comment by Samyak2** [comment]
Just noticed that Tokenizer v0.2.7 was released with my fix! I'll update this PR soon. Thanks for your patience :smile:

File: http-ui/src/main.rs [hunk 1]
```diff
@@ -160,13 +161,21 @@ impl<'a, A: AsRef<[u8]>> Highlighter<'a, A> {
...
+                            string.push_str("<mark>");
+                            string.push_str(
+                                chars.take(chars_to_highlight).collect::<String>().as_str(),
+                            );
+                            string.push_str("</mark>");
+
+                            let chars = graphemes.into_iter().skip(chars_to_highlight);
+                            string.push_str(chars.collect::<String>().as_str());
+                        } else {
+                            string.push_str(word);
```
[the remark this answers is not in the document]
Comment (Samyak2) [reply http-ui/src/main.rs]: Done in c0313f3

**Comment by Samyak2** [comment]
The highlighting works well now! Screenshots:

Spectacle_last

Spectacle_last

Spectacle_last

---

But, there is an issue if only a part of `briefcase` matches. For example, `gobrifcaseod` does not highlight the last `d` since only the first 12 chars match. `gobrifcaseood` matches the whole word though.

I think it's a minor issues and it works quite well even with it.

Spectacle_last

**Comment by ManyTheFish** [comment]
Hello @Samyak2, I agree with you, despite this minor issue It's a big improvement, thanks a lot for your work!

Bors merge

**Review by ManyTheFish — APPROVED** [review]

**Comment by Samyak2** [comment]
Thank you for guiding me through all of my PRs :heart:

**Pull request merged**